### PR TITLE
Remove ops from Layer Creator/ Node Converter - part 2

### PR DESCRIPTION
**Pull request by blesniewski**
Removing ops from layer creator:
- power
- prelu
- tile
- relu
- selu
- softmax
- tanh
- split
- reshape
- reverse sequence
- proposal
- prior box
- roi align
- prior box clustered
41390

### CODE REVIEW

**Comment by blesniewski** [comment]
So far this is just to double-check that all tests are passing, I'll be adding more reviewers when the changes are ready.

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp [hunk 1]
```diff
@@ -505,11 +492,17 @@ std::shared_ptr<ngraph::Node> V10Parser::createNode(const std::vector<ngraph::Ou
     if (!ngraphNode && opsets.count(params.version)) {
         auto opset = opsets.at(params.version);
 
-        if (!opset.contains_type(params.type)) {
+        // ROIPooling was missing in opset1 and was added in opset2
+        if (params.type == "ROIPooling" && params.version == "opset1") {
+            opset = opsets.at("opset2");
+        }
+
+        if (!opset.contains_type_insensitive(params.type)) {
```
Comment (sdurawa) [inline inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:500]: which op requires insensitive check?

Comment (blesniewski) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:500]: In this case it was PReLU

Comment (sdurawa) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:500]: Looks like there is a mistake in type name in functional test. I checked open model zoo example (https://download.01.org/opencv/2021/openvinotoolkit/2021.1/open_model_zoo/models_bin/2/person-reidentification-retail-0287/FP32/person-reidentification-retail-0287.xml) and type name for v10 is "PReLU".

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp [hunk 2]
```diff
@@ -505,11 +492,17 @@ std::shared_ptr<ngraph::Node> V10Parser::createNode(const std::vector<ngraph::Ou
     if (!ngraphNode && opsets.count(params.version)) {
         auto opset = opsets.at(params.version);
 
-        if (!opset.contains_type(params.type)) {
+        // ROIPooling was missing in opset1 and was added in opset2
```
Comment (ilyachur) [inline inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:495]: Am I right that MO generates IR where `ROIPooling` exists in the `opset1`?

@lazarevevgeny I think we should fix it on the MO side if MO generates `opset1` for `ROIPooling`

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 1 above)
Comment (ilyachur) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:500]: It is a correct change. Because nGraph op type is `Relu` but in the MO and in the spec we use `ReLU`.
We have some problems with case sensitive.

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 2 above)
Comment (blesniewski) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:495]: In the ngraph reader IE Functional tests we have IRs with ROIPooling in opset1.
There is also a comment in opset2_tbl file that says that this op should be available in opset1 but was not by mistake, hence my comment here and this fix.

Comment (ilyachur) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:495]: I think that we can change the version in the test, instead of this WA. But also need to check that MO generates the right version.

Comment (lazarevevgeny) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:495]: I checked that MO from 2020.3 generated ROIPooling with opset1. Starting from 2020.4 MO generates ROIPooling with opset2.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 3]
```diff
@@ -569,6 +577,218 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+            details::convertPrecision(node->get_output_element_type(0))};
+        auto parseBoolStrToIntStr = [](const std::string &param) -> const std::string {
+            if (param == "true") {
+                return "1";
+            }
+            else if (param == "false") {
+                return "0";
+            }
+            return param;
+        };
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:621]: there should be similar function already. 

@ilyachur can you provide it?

Comment (lazarevevgeny) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:621]: Even if it does not exist then it should be placed to some common place.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 4]
```diff
@@ -569,6 +577,218 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+        LayerParams attrs = {node->get_friendly_name(), "PriorBoxClustered",
+            details::convertPrecision(node->get_output_element_type(0))};
+        auto res = std::make_shared<CNNLayer>(attrs);
+        res->params = params;
+        if (res->params["clip"] == "true") {
+            res->params["clip"] = "1";
+        }
+        else if (res->params["clip"] == "false") {
+            res->params["clip"] = "0";
+        }
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:671]: common function like in PriorBoxIE could be used.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 5]
```diff
@@ -569,6 +577,218 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+        return res;
+    });
+
+    addSpecificCreator({"ProposalIE"},
+                       [](const std::shared_ptr<::ngraph::Node>& node,
+                          const std::map<std::string, std::string>& params) -> CNNLayerPtr {
+        LayerParams attrs = {node->get_friendly_name(), "Proposal",
+            details::convertPrecision(node->get_output_element_type(0))};
+        auto parseBoolStrToIntStr = [](const std::string &param) -> const std::string {
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:685]: The same function like in lines 613-621

File: inference-engine/src/legacy_api/src/ngraph_ops/prior_box_clustered_ie.cpp [hunk 6]
```diff
@@ -37,3 +37,27 @@ std::shared_ptr<Node> op::PriorBoxClusteredIE::clone_with_new_inputs(const Outpu
...
+    visitor.on_attribute("step_h", m_attrs.step_heights);
+    if(step != 0) {
+        // deserialization: if step_w/h is 0 replace it with step
+        if (m_attrs.step_widths == 0) {
+            m_attrs.step_widths = step;
+        }
+        if (m_attrs.step_heights == 0) {
+            m_attrs.step_heights = step;
+        }
+    }
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/ngraph_ops/prior_box_clustered_ie.cpp:56]: Could you explain why this code is needed?

File: inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp [hunk 7]
```diff
@@ -159,6 +159,8 @@ namespace FuncTestUtils {
                 }
             } else {
                 if (item.first == "originalLayersNames") continue;
+                // pooled_h and pooled_w are output_size for ROIPooling
```
Comment (lazarevevgeny) [inline inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp:162]: The comment does not explain why the code was modified. Could you provide more detailed explanation?

File: ngraph/core/src/op/prior_box_clustered.cpp [hunk 8]
```diff
@@ -96,13 +96,31 @@ shared_ptr<Node> op::PriorBoxClustered::clone_with_new_inputs(const OutputVector
...
+        // deserialization:
+        // if step_w/h is 0 or did not change, replace it with step
+        if (m_attrs.step_widths == 0 || m_attrs.step_widths == step_w_tmp)
+        {
+            m_attrs.step_widths = step;
+        }
+        if (m_attrs.step_heights == 0 || m_attrs.step_heights == step_h_tmp)
+        {
+            m_attrs.step_heights = step;
+        }
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/prior_box_clustered.cpp:117]: Why this logic is in the visit attributes?

File: ngraph/python/tests/test_ngraph/test_create_op.py [hunk 9]
```diff
@@ -896,8 +896,8 @@ def test_prior_box_clustered(int_dtype, fp_dtype):
     image_size = np.array([64, 64], dtype=int_dtype)
     attributes = {
         "offset": fp_dtype(0.5),
-        "widths": np.array([4.0, 2.0, 3.2], dtype=fp_dtype),
-        "heights": np.array([1.0, 2.0, 1.0], dtype=fp_dtype),
+        "width": np.array([4.0, 2.0, 3.2], dtype=fp_dtype),
```
Comment (lazarevevgeny) [inline ngraph/python/tests/test_ngraph/test_create_op.py:899]: this looks like not backward compatible change. If we want to make it then we need to introduce new version of the operation.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 3 above)
Comment (ilyachur) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:621]: Such method doesn't exist. But from my point of view we can add something like: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/legacy_api/include/legacy/ie_layers.h#L336

File: inference-engine/src/legacy_api/src/ngraph_ops/prior_box_clustered_ie.cpp (hunk 6 above)
Comment (ilyachur) [reply inference-engine/src/legacy_api/src/ngraph_ops/prior_box_clustered_ie.cpp:56]: Maybe it will be better to discuss offline. In general, in the spec and in the MO we have a logic which depends on the step attribute.
And we need this code in order to emulate the same behavior

File: inference-engine/src/readers/ir_reader/ie_ir_parser.cpp (hunk 2 above)
Comment (ilyachur) [reply inference-engine/src/readers/ir_reader/ie_ir_parser.cpp:495]: Hmm, ok, it means that we need this code, because 2020.3 is LTS release.

File: ngraph/core/src/op/prior_box_clustered.cpp (hunk 8 above)
Comment (blesniewski) [reply ngraph/core/src/op/prior_box_clustered.cpp:117]: So this op can have three parameters that relate to each other: _step_, _step_w_, _step_h_.
The IR can contain 1(_step_) or 2(_step_w_, _step_h_) or 3(_step_, _step_w_, _step_h_) of them.

- If the IR has _step_w_ and _step_h_ specified we use that. 
- If _step_w_ or _step_h_ is 0 we replace it with the value of step.
- If _step_ = 0 and _step_w_==_step_h_ we assign _step_w's_ value to step.
- Only step is present, assign its value to _step_w_ and _step_h_.

This logic is fairly easy to do, just not in visit attributes, where we have to make a universal solution for all uses of this method (due to the nature of this pattern, we don't know if _visit_attributes_ is being run on serialization/deserialization, or conversion to CNNN path).

So without this logic, when we're reading the IR and we have only step parameter available in the IR, we would not initialize _step_w_ and _step_h_, which would give the wrong result as the op would have wrong parameters.

This could all be made much more clear if there was a way to distinguish the path we're on or overload _visit_attributes_ method for _XMLDeserializer_, but AFAIK it's not possible.

File: ngraph/python/tests/test_ngraph/test_create_op.py (hunk 9 above)
Comment (blesniewski) [reply ngraph/python/tests/test_ngraph/test_create_op.py:899]: So what can I do here, if _visit_attributes_ needs this to be stated explicitly if the parameter's name is height or heights, and the spec and IR in IE functional tests say this should be _width_ and _height_?

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 3 above)
Comment (blesniewski) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp]: I agree on this, I've just added such function, but I'm not sure about it's placement, can you take a look?

File: inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp (hunk 7 above)
Comment (blesniewski) [reply inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp:162]: Sure, the specification says that there should be two parameters- _pooled_h_ and _pooled_w_, our implementation of this op has a single parameter _output_size_, we shouldn't be failing just if this parameter is not present.

You can take a look at the single commit I've added just for _roi pooling_ to maybe get a better perspective on what changes I made to this operator.

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 3 above)
Comment (ilyachur) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp]: I think you can extend CNNLayer methods. It is a legacy API and we use it only to convert Ngraph to legacy API

Comment (blesniewski) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp]: Added a CNNLayer method for rhat, I'm going to resolve this conversation

File: inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp (hunk 7 above)
Comment (lazarevevgeny) [reply inference-engine/tests/ie_test_utils/functional_test_utils/network_utils.cpp:162]: Please, add this comment to the code itself :)

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 10]
```diff
@@ -538,6 +538,196 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+
+            // mxnet-like PriorBox
+            auto img_H = img_shape[2];
+            auto data_H = data_shape[2];
+
+            auto step = std::stof(res->params["step"]);
+            if (step == -1)
+                step = static_cast<float>(1. * img_H / data_H);
+            else
+                step *= 1. * img_H;
```
Comment (ilyachur) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:596]: Why do we need this expression `1. * img_H`?
Can we simplify it to
```suggestion
                step *= img_H;
```

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 11]
```diff
@@ -538,6 +538,196 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+            auto img_shape = img_pshape.to_shape();
+            if (img_shape.size() != 4) THROW_IE_EXCEPTION << "PriorBox has " << data_shape.size() << " items in 1-port input, 4 expected";
+
+            // mxnet-like PriorBox
+            auto img_H = img_shape[2];
+            auto data_H = data_shape[2];
+
+            auto step = std::stof(res->params["step"]);
+            if (step == -1)
+                step = static_cast<float>(1. * img_H / data_H);
```
Comment (ilyachur) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:594]: Can we simplify it to:
```suggestion
                step = img_H / static_cast<float>(data_H);
```

File: ngraph/core/include/ngraph/op/prior_box.hpp [hunk 12]
```diff
@@ -44,7 +44,7 @@ namespace ngraph
             float step = 1.0f;
```
Comment (sdurawa) [inline ngraph/core/include/ngraph/op/prior_box.hpp:44]: According to spec (https://docs.openvinotoolkit.org/latest/openvino_docs_ops_detection_PriorBox_1.html) step shall be initialized with 0.

**Review by sdurawa — APPROVED** [review]

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 13]
```diff
@@ -538,6 +538,196 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+                       [](const std::shared_ptr<::ngraph::Node>& node,
+                          const std::map<std::string, std::string>& params) -> CNNLayerPtr {
+        LayerParams attrs = {node->get_friendly_name(), "ReverseSequence",
+            details::convertPrecision(node->get_output_element_type(0))};
+        auto res = std::make_shared<ReverseSequenceLayer>(attrs);
+        res->params = params;
+        return res;
+    });
+
+    addSpecificCreator({"ROIPooling"},
```
Comment (ilyachur) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:670]: I don't think that we need specific creator for ROIPooling. We don't change the type or any attributes, it should work from the box.

**Review by ilyachur — APPROVED** [review]

**Comment by blesniewski** [comment]
@lazarevevgeny Could you take a look please?

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp [hunk 14]
```diff
@@ -538,6 +538,186 @@ InferenceEngine::details::CNNLayerCreator::CNNLayerCreator(const std::shared_ptr
...
+    });
+
+    addSpecificCreator({"Reshape"},
+                       [](const std::shared_ptr<::ngraph::Node>& node,
+                          const std::map<std::string, std::string>& params) -> CNNLayerPtr {
+        LayerParams attrs = {node->get_friendly_name(), "Reshape",
+            details::convertPrecision(node->get_output_element_type(0))};
+        auto res = std::make_shared<ReshapeLayer>(attrs);
+        return res;
+    });
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:658]: Do we really need these 2 creators? The same comment for Softmax

File: inference-engine/src/legacy_api/src/ie_layers.cpp [hunk 15]
```diff
@@ -306,6 +306,16 @@ std::string CNNLayer::GetParamAsString(const char* param) const {
     return (*it).second;
 }
 
+std::string CNNLayer::getBoolStrParamAsIntStr(const char *param) const {
+    std::string val = GetParamAsString(param);
+    if (val == "true") {
```
Comment (lazarevevgeny) [inline inference-engine/src/legacy_api/src/ie_layers.cpp:311]: Please, add check for "True" and "False" also. Or just convert the string to lowercase always.

@ilyachur , is it acceptable?

File: inference-engine/src/plugin_api/debug.h [hunk 16]
```diff
@@ -207,5 +207,6 @@ inline std::string tolower(const std::string& s) {
     std::transform(s.begin(), s.end(), ret.begin(), ::tolower);
     return ret;
 }
+
```
Comment (lazarevevgeny) [inline inference-engine/src/plugin_api/debug.h:210]: Is this change really needed?

File: inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp (hunk 14 above)
Comment (ilyachur) [reply inference-engine/src/legacy_api/src/convert_function_to_cnn_network.cpp:658]: We need to have these creators, because we create specific CNNLayers (Reshape layer in this case).

File: inference-engine/src/legacy_api/src/ie_layers.cpp (hunk 15 above)
Comment (ilyachur) [reply inference-engine/src/legacy_api/src/ie_layers.cpp:311]: Yes I think it is acceptable.

Comment (blesniewski) [reply inference-engine/src/legacy_api/src/ie_layers.cpp:311]: I will add it, but I'd like to see any example where we could have 'True' or 'False' in CNNLayer's params.

File: inference-engine/src/plugin_api/debug.h (hunk 16 above)
Comment (blesniewski) [reply inference-engine/src/plugin_api/debug.h:210]: No, I'm removing it.

File: inference-engine/src/legacy_api/src/ie_layers.cpp (hunk 15 above)
Comment (lazarevevgeny) [reply inference-engine/src/legacy_api/src/ie_layers.cpp]: Previously MO generated IRs where boolean value could be string "True" or "False". This is why I asked to make this change. Glad to hear that you will change this

**Review by lazarevevgeny — APPROVED** [review]
